src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,365 @@
+recompiled.
+
+We could do something like split up `CustomPrelude.hs` into a module that is
+depended upon by `A` and `B` and another that is depended upon by `C` and `D`.
+This means that changes in `PreludeAB.hs` now only recompile `A` and `B` and the
+same for `PreludeCD.hs` and `C` and `D`. In this way we reduce overall compile
+times by making it clear to GHC, at the module level, which code depends on
+what. GHC will use this information to see what work it has already done that is
+still valid, and what needs to be re-computed.
+
+```
+
+                                       Main.hs
+                                      /   |   \
+                                     /   / \   \
+                                    A   B   C   D
+                                   /   /     \   \
+                              PreludeAB.hs   PreludeCD.hs
+
+```
+
+> Note: Main.hs has to be recompiled every time, regardless, so it does help if
+> you can move code that doesn't change frequently down the dependency graph.
+
+For a deeper exposition of module recompilation the
+[GHC docs](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/recompilation-avoidance)
+are a great resource.
+
+A crucial point to understand here is that `GHC` is responsible for this
+recompilation logic, and a modules capabilities are defined in the Haskell
+language specification. Why, as cabal users, is it important for us to
+understand the module dependency graph, and how to use modules in general?
+
+The answer is two-fold:
+
+1. We will often interface with GHC's recompilation functionality through
+   `cabal` with commands like `cabal build` and `cabal repl`.
+2. The module is the smallest unit of concern within cabal, and therefore it is
+   worthwhile to understand its meaning, use cases, and impact on our codebase.
+
+## What is a Component?
+
+If modules are the construct that allow us to manage the internal dependencies
+of our project, then components are the first glimpse of the external.
+Components aggregate all of our internal dependencies to be exposed external,
+and components also declare all of the external packages that we depend on so
+that they can be made available as modules within our code.
+
+When arranging components Cabal needs to be aware of all the modules that we
+would like to expose externally. Now _what_ we choose to expose is up to us.
+Sometimes we just want to write an application, in which case we don't need to
+expose any modules, just the `main` entrypoint that gets called when we run our
+executable. Sometimes we want to write a library for others to use, and we want
+to liberally expose everything. Sometimes we want to write a library with some
+nasty internal machinery which we would prefer to keep hidden, and only expose
+the clean external api. All of these use cases can be expressed using
+components.
+
+Components are represented by top level keys within a `.cabal` file. These are
```

Comment:
This is a good place to mention that each kind of component can appear multiple times in .cabal file and can be named (and that libraries don't have to be named). And then ignore the topic of multiple private and public libraries completely, perhaps giving a link to a chapter about multiple private and public libraries (multiple exes, etc. are trivial)